File: trunk/dcget/src/peer/PeerConnection.java
package peer;

import java.net.Socket;
import logger.ILogger;

public class PeerConnection {

    private ILogger logger;
    private IPeerEventHandler handler;
    private PeerReader reader;
    private PeerWriter writer;

    public PeerConnection(ILogger logger, IPeerEventHandler handler, String ip, int port) throws Exception {
        this.logger = new PeerLogger(logger, ip);
        this.handler = handler;
        connect(ip, port);
    }

    public void run() throws Exception {
        reader.read();
    }

    private void connect(String ip, int port) throws Exception {
        Socket sock = new Socket(ip, port);
        reader = new PeerReader(sock.getInputStream(), logger);
        writer = new PeerWriter(sock.getOutputStream(), logger);
        handler.onPeerConnected(this);
    }

    public void handshake(String nick) throws Exception {
        writer.sendMyNick(nick);
        writer.sendLock("some_random_lock", "kio_dcpp");
    }

}

File: dcget++/PeerConnection.cpp
#include "PeerConnection.h"

#include "MyNickHandler.h"
#include "FileLengthHandler.h"
#include "LockHandler.h"
#include "KeyHandler.h"
#include "PeerLockHandler.h"
#include "ErrorHandler.h"
#include "MaxedOutHandler.h"
#include "DataHandler.h"

void PeerConnection::connect(const std::string& ip, int port) {
    logger->debug("connecting to " + ip);
    sock = new Socket();
    sock->connect(ip, port);
    sock->set_non_blocking(true);
    logger->debug("connected");
    reader = new PeerReader(sock, logger);
    reader->registerHandler(new MyNickHandler(handler, this));
    reader->registerHandler(new FileLengthHandler(this, handler));
    reader->registerHandler(new PeerLockHandler(this));
    //reader->registerHandler(new DirectionHandler(handler, this));
    reader->registerHandler(new KeyHandler(this, handler));
    reader->registerHandler(new ErrorHandler(this, handler));
    reader->registerHandler(new MaxedOutHandler(this, handler));
    reader->registerHandler(new DataHandler(this, handler));
    writer = new PeerWriter(sock, logger);
    handler->onPeerConnected(this);
}

void PeerConnection::onKeyReceived() {
    handler->onHandShakeDone(this);
}